zetaclient/chains/ton/observer/outbound.go

```diff
@@ -0,0 +1,323 @@
+
+		return nil
+	}
+
+	return ticker.Run(
+		ctx,
+		initialInterval,
+		task,
+		ticker.WithStopChan(ob.StopChannel()),
+		ticker.WithLogger(ob.Logger().Outbound, "WatchOutbound"),
+	)
+}
+
+// observeOutboundTrackers pulls outbounds trackers from zetacore,
+// fetches txs from TON and stores them in memory for further use.
+func (ob *Observer) observeOutboundTrackers(ctx context.Context) error {
+	var (
+		chainID  = ob.Chain().ChainId
+		zetacore = ob.ZetacoreClient()
+	)
+
+	trackers, err := zetacore.GetAllOutboundTrackerByChain(ctx, chainID, interfaces.Ascending)
+	if err != nil {
+		return errors.Wrap(err, "unable to get outbound trackers")
+	}
+
+	for _, tracker := range trackers {
+		nonce := tracker.Nonce
+
+		// If outbound is already in memory, skip.
+		if _, ok := ob.getOutboundByNonce(nonce); ok {
+			continue
+		}
+
+		// Let's not block other cctxs from being processed
+		cctx, err := zetacore.GetCctxByNonce(ctx, chainID, nonce)
+		if err != nil {
+			ob.Logger().Outbound.
+				Error().Err(err).
+				Uint64("outbound.nonce", nonce).
+				Msg("Unable to get cctx by nonce")
+
+			continue
+		}
+
+		for _, txHash := range tracker.HashList {
+			if err := ob.processOutboundTracker(ctx, cctx, txHash.TxHash); err != nil {
+				ob.Logger().Outbound.
+					Error().Err(err).
+					Uint64("outbound.nonce", nonce).
+					Str("outbound.hash", txHash.TxHash).
+					Msg("Unable to check transaction by nonce")
+			}
+		}
+	}
+
+	return nil
+}
+
+// processOutboundTracker checks TON tx and stores it in memory for further processing
+// by VoteOutboundIfConfirmed.
+func (ob *Observer) processOutboundTracker(ctx context.Context, cctx *cc.CrossChainTx, txHash string) error {
+	if cctx.InboundParams.CoinType != coin.CoinType_Gas {
+		return errors.New("only gas cctxs are supported")
+	}
+
+	lt, hash, err := liteapi.TransactionHashFromString(txHash)
+	if err != nil {
+		return errors.Wrap(err, "unable to parse tx hash")
+	}
+
+	rawTX, err := ob.client.GetTransaction(ctx, ob.gateway.AccountID(), lt, hash)
+	if err != nil {
+		return errors.Wrap(err, "unable to get transaction form liteapi")
+	}
+
+	tx, err := ob.gateway.ParseTransaction(*rawTX)
+	if err != nil {
+		return errors.Wrap(err, "unable to parse transaction")
+	}
+
+	receiveStatus, err := ob.determineReceiveStatus(tx)
+	if err != nil {
+		return errors.Wrap(err, "unable to determine outbound outcome")
+	}
+
+	// TODO: Add compliance check
+	// https://github.com/zeta-chain/node/issues/2916
+
+	nonce := cctx.GetCurrentOutboundParam().TssNonce
+	ob.setOutboundByNonce(outbound{tx, receiveStatus, nonce})
+
+	return nil
+}
+
+func (ob *Observer) determineReceiveStatus(tx *toncontracts.Transaction) (chains.ReceiveStatus, error) {
+	_, evmSigner, err := extractWithdrawal(tx)
+	switch {
+	case err != nil:
+		return 0, err
+	case evmSigner != ob.TSS().EVMAddress():
+		return 0, errors.New("withdrawal signer is not TSS")
+	case !tx.IsSuccess():
+		return chains.ReceiveStatus_failed, nil
```

Comment:
> Can a failed tx effectively increase gateway nonce?

Not in the current implementation. Just placed this part for the sake of completeness. It might fail if we introduce net-new logic in the gw that might fail in some edge-case.

Reply:
https://github.com/zeta-chain/node/pull/3020/commits/f96229d526764543de91995e3b95d3afcaaa2ef3